### PR TITLE
Add timeout flag to bootstrap

### PR DESCRIPTION
**Pull request by fejta**
ref #1226 #1225

Add a `--timeout` flag to bootstrap.py and corresponding unit tests.
Defaults to 0 and therefore does nothing.
Will migrate some jobs in subsequent PRs.

This will allow us to remove the `### Reporting` boilerplate in `job/foo.sh`

### CODE REVIEW

**Comment by fejta** [comment]
The failure is https://github.com/kubernetes/test-infra/pull/1469

**Comment by fejta** [comment]
@k8s-bot bazel test this

File: jenkins/bootstrap.py [hunk 1]
```diff
@@ -55,45 +56,91 @@
...
+    return True
+
+
+def _call(end, cmd, stdin=None, check=True, output=None):
     """Start a subprocess."""
     logging.info('Call:  %s', ' '.join(pipes.quote(c) for c in cmd))
+    begin = time.time()
+    if end:
+        end = max(end, time.time() + 60)  # Allow at least 60s per command
+        logging.info('Limiting call to %.1f minutes', elapsed(begin))
```
Comment (krzyzacy) [inline jenkins/bootstrap.py]: do you want to print `(end-begin)/60` here?

File: jenkins/bootstrap.py [hunk 2]
```diff
@@ -55,45 +56,91 @@
...
 
 
-def call(cmd, stdin=None, check=True, output=None):
+def read_all(end, stream, append):
+    """Read all buffered lines from a stream."""
+    while not end or time.time() < end:
+        line = stream.readline()
+        if not line:
+            return True  # Read everything
+        append(line[:-1])  # Ignore the \n at the end
```
Comment (krzyzacy) [inline jenkins/bootstrap.py]: why ignore `\n`?

File: jenkins/bootstrap.py [hunk 3]
```diff
@@ -55,45 +56,91 @@
...
+    if timeout:
+        code = code or 124
+        logging.error('Build timed out')
+    if code:
+        logging.error('Build failed')
+    logging.info(
+        'process %d exited with code %d after %.1fm',
+        proc.pid, code, elapsed(begin))
+    out.append('')
     lines = output and '\n'.join(out)
```
Comment (krzyzacy) [inline jenkins/bootstrap.py]: I think ignore `\n` is correct, thus it was wrong previously(`out.append(line)`)?

File: jenkins/bootstrap.py [hunk 4]
```diff
@@ -374,7 +424,7 @@ def node():
     return os.environ[NODE_ENV]
 
 
-def find_version():
+def find_version(call):
```
Comment (krzyzacy) [inline jenkins/bootstrap.py]: can you explain how `call` is being used here inside the function? Do you need some implementation as above in  `checkout`?

File: jenkins/bootstrap_test.py [hunk 5]
```diff
@@ -90,44 +92,140 @@ def Bomb(*a, **kw):
...
+        """Note we reached the end of the stream."""
+        lines = []
+        self.readline = lambda: ''
+        with Stub(select, 'select', lambda *a, **kw: ([self],[],[])):
+            done = bootstrap.read_all(self.endless, self, lines.append)
+
+        self.assertTrue(done)
+
+
+
```
Comment (krzyzacy) [inline jenkins/bootstrap_test.py]: nit: one less empty line here

File: jenkins/bootstrap_test.py [hunk 6]
```diff
@@ -90,44 +92,140 @@ def Bomb(*a, **kw):
...
+    def testTerminate_Kill(self):
+        """Kill process group if after end and kill is true."""
+        with Stub(os, 'getpgid', self.getpgid), Stub(os, 'killpg', self.killpg):
+            timeout = bootstrap.terminate(time.time() - 50, self, True)
+        self.assertTrue(timeout)
+        self.assertFalse(self.terminated)
+        self.assertTrue(self.killed)
+        self.assertEquals(self.pid, self.got)
+        self.assertEquals(self.killed_pg, (self.pgid, signal.SIGKILL))
+
```
Comment (krzyzacy) [inline jenkins/bootstrap_test.py]: nit: one more empty line?

File: jenkins/bootstrap.py (hunk 1 above)
Comment (fejta) [inline jenkins/bootstrap.py]: Done

File: jenkins/bootstrap.py (hunk 3 above)
Comment (fejta) [inline jenkins/bootstrap.py]: Ack

File: jenkins/bootstrap.py (hunk 4 above)
Comment (fejta) [inline jenkins/bootstrap.py]: This is hidden by the review but it is used in [L400](https://github.com/kubernetes/test-infra/blob/master/jenkins/bootstrap.py#L400) which is now L450 in the right side.

File: jenkins/bootstrap_test.py (hunk 5 above)
Comment (fejta) [inline jenkins/bootstrap_test.py]: Done

File: jenkins/bootstrap_test.py (hunk 6 above)
Comment (fejta) [inline jenkins/bootstrap_test.py]: Done

File: jenkins/bootstrap.py (hunk 4 above)
Comment (krzyzacy) [inline jenkins/bootstrap.py]: got it, thanks

**Comment by krzyzacy** [comment]
lgtm, guess give @spxtr a chance to take another look

**Pull request merged**